opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -367,7 +366,38 @@ namespace Opm {
 
         this->resetWGState();
 
-        const int reportStepIdx = simulator_.episodeIndex();
+        if (lastStepFailed) {
+            // if wells are recently shut we need to do the
+            // full beginTimeStep()
+            bool any_closed_wells_this_step = false;
+            const int nw = this->numLocalWells();
+            int w = 0;
+            while (!any_closed_wells_this_step && w < nw) {
+                const Well& well_ecl = this->wells_ecl_[w];
+                const std::string& well_name = well_ecl.name();
+                if (this->wellTestState().well_is_closed(well_name)) {
+                    const bool closed_this_step = (this->wellTestState().lastTestTime(well_name) == simulator_.time());
+                    any_closed_wells_this_step = closed_this_step;
+                }
+                w += 1;
+            }
```

Comment:
Think you need to globally communicate any_closed_wells_this_step (preferably an int equivalent?) to ensure all procs do the same thing.

Reply:
Thanks for noticing.